spdx-utils/src/main/kotlin/model/SpdxExternalReference.kt

```diff
@@ -56,25 +63,45 @@ data class SpdxExternalReference(
         OTHER
     }
 
-    // TODO: Using new values as per https://github.com/spdx/spdx-spec/issues/451.
-    enum class Type(
-        val typeName: String,
+    sealed class Type(
+        @JsonValue
+        val name: String,
         val category: Category
     ) {
-        BOWER("bower", Category.PACKAGE_MANAGER),
-        CPE_22_TYPE("cpe22Type", Category.SECURITY),
-        CPE_23_TYPE("cpe23Type", Category.SECURITY),
-        MAVEN_CENTRAL("maven-central", Category.PACKAGE_MANAGER),
-        NPM("npm", Category.PACKAGE_MANAGER),
-        NUGET("nuget", Category.PACKAGE_MANAGER),
-        PURL("purl", Category.PACKAGE_MANAGER),
-        SWH("swh", Category.PERSISTENT_ID),
-        OTHER("OTHER", Category.OTHER);
+        object Cpe22Type : Type("cpe22Type", Category.SECURITY)
+        object Cpe23Type : Type("cpe23Type", Category.SECURITY)
+
+        object Bower : Type("bower", Category.PACKAGE_MANAGER)
+        object MavenCentral : Type("maven-central", Category.PACKAGE_MANAGER)
+        object Npm : Type("npm", Category.PACKAGE_MANAGER)
+        object NuGet : Type("nuget", Category.PACKAGE_MANAGER)
+        object Purl : Type("purl", Category.PACKAGE_MANAGER)
+
+        object SoftwareHeritage : Type("swh", Category.PERSISTENT_ID)
+
+        class Other(typeName: String) : Type(typeName, Category.OTHER)
     }
 
     init {
         require(referenceLocator.isNotBlank()) { "The referenceLocator must not be blank." }
 
-        require(referenceType.isNotBlank()) { "The referenceType must not be blank." }
+        if (referenceType.category != Category.OTHER) {
+            require(referenceType.category == referenceCategory) {
```

Comment:
Could remove one level of nesting by: `require(referenceType.category == Category.OTHER || referenceType.category == referenceCategory)`

Reply:
If `referenceType.category` is `Category.OTHER`, why must then `referenceCategory` not be `Category.OTHER` as well?

Reply:
I wasn't sure how the read the [specification for this](https://spdx.github.io/spdx-spec/appendix-VI-external-repository-identifiers/), but my thinking was that if a user wants to refer to package manager unknown to SPDX, would make sense to allow the user to still use `PACKAGE_MANAGER` as the category, even if our "auto-detecting" of the category would say `OTHER`.

Reply:
Thanks, that makes sense.